### PR TITLE
amp-script: Refactor and improve sanitization coverage

### PR DESCRIPTION
**Pull request by choumx**
~Requires https://github.com/ampproject/worker-dom/pull/420 to be released first.~

~@dvoytenko Will address the attribute removal issue and setting up custom hooks for `isValidAttribute` soon.~

- Add private `SanitizerImpl` class to amp-script.js that encapsulates validation logic.
- Upgrades `@ampproject/worker-dom` to 0.10.0.
- Use new `mutateAttribute()` & `mutateProperty()` APIs from https://github.com/ampproject/worker-dom/pull/420.
-  Applies AMP runtime validation rules with new `purifier.js#validateAttributeChange()` with unit tests.
- Adds integration test that checks that `script` and `img` insertions are sanitized.

### CODE REVIEW

File: extensions/amp-script/0.1/amp-script.js [hunk 1]
```diff
@@ -131,7 +125,45 @@ export class AmpScript extends AMP.BaseElement {
       onReceiveMessage: data => {
         dev().info(TAG, 'From worker:', data);
       },
-    }).then(workerDom => {
+      onMutationPump: this.mutationPump_.bind(this),
```
Comment (dvoytenko) [inline extensions/amp-script/0.1/amp-script.js]: Hmm. I'm pretty sure the naming of these has changed.

File: extensions/amp-script/0.1/amp-script.js [hunk 2]
```diff
@@ -131,7 +125,45 @@ export class AmpScript extends AMP.BaseElement {
...
+      onLongTask: promise => {
+        this.userActivation_.expandLongTask(promise);
+        // TODO(dvoytenko): consider additional "progress" UI.
+      },
+      sanitizer: {
+        sanitize: node => {
+          // DOMPurify sanitizes unsafe nodes by detaching them from parents.
+          // So, an unsafe `node` that has no parent will cause a runtime error.
+          // To avoid this, wrap `node` in a <div> if it has no parent.
+          const useWrapper = !node.parentNode;
```
Comment (dvoytenko) [inline extensions/amp-script/0.1/amp-script.js]: IMHO, this is a lot of code to inline...

File: extensions/amp-script/0.1/amp-script.js (hunk 1 above)
Comment (choumx) [reply extensions/amp-script/0.1/amp-script.js]: Bad merge, removed.

File: extensions/amp-script/0.1/amp-script.js (hunk 2 above)
Comment (choumx) [reply extensions/amp-script/0.1/amp-script.js]: Good point. Encapsulated into a `SanitizerImpl` private class.

File: extensions/amp-script/0.1/amp-script.js [hunk 3]
```diff
@@ -131,7 +119,10 @@ export class AmpScript extends AMP.BaseElement {
       onReceiveMessage: data => {
         dev().info(TAG, 'From worker:', data);
       },
-    }).then(workerDom => {
+      sanitizer: new SanitizerImpl(this.win),
```
Comment (dvoytenko) [inline extensions/amp-script/0.1/amp-script.js]: nit: bring this up with other properties (above callbacks).

File: extensions/amp-script/0.1/amp-script.js [hunk 4]
```diff
@@ -184,6 +175,105 @@ export class AmpScript extends AMP.BaseElement {
...
+    return true;
+  }
+
+  /**
+   * @param {!Node} node
+   * @param {string} attribute
+   * @param {string} value
+   * @return {boolean}
+   */
+  mutateAttribute(node, attribute, value) {
```
Comment (dvoytenko) [inline extensions/amp-script/0.1/amp-script.js]: This seems like enough code to warrant a separate test or two?

Comment (choumx) [reply extensions/amp-script/0.1/amp-script.js]: Done.

File: extensions/amp-script/0.1/amp-script.js (hunk 3 above)
Comment (choumx) [reply extensions/amp-script/0.1/amp-script.js]: Done.

**Pull request merged**